### PR TITLE
Test generator for output consistency checks (V1)

### PR DESCRIPTION
**Pull request by nrainer-materialize**
This closes https://github.com/MaterializeInc/materialize/issues/19438. That is part of https://github.com/MaterializeInc/materialize/issues/18858.

### Overview
Description of concept: https://www.notion.so/materialize/Extensive-Output-Consistency-Tests-fd1ec02f5f604e6ba6e310a4fba96ced.

### Motivation
This PR adds a known-desirable feature.

### Tips for reviewer
* Start with the [README.md](https://github.com/nrainer-materialize/materialize/blob/feature/18858-extensive-output-consistency/test/output-consistency/README.md)

* To view the generated queries, run `bin/output-consistency --max-runtime-in-sec 2 --verbose`. To avoid bloating the output with single-column queries expected to result in database errors, consider adding `--avoid-expressions-expecting-db-error` to skip those.

#### Basic functionality
* ✅ setup data structures
* ✅ define some data types, functions, and other expressions
* ✅ generate statements to set up and populate tables and views
* ✅ generate queries
* ✅ execute queries and collect results
* ✅ validate the outcome (result and error) of executed queries
* ✅ support functions with arguments of different data types
* ✅ generate expressions for specifically for operations
* ✅ allow skipping known output consistency deviations

#### Execution
* ✅ setup mzcompose entry point
* ✅ integrate into Buildkite
* ✅ allow stand-alone execution without mzcompose
* ✅ provide dry-run mode

#### Optimizations
* ✅ identify and separate expressions likely causing a query error
* ✅ split query and retry execution if outcome is database error and multiple columns are involved
* ✅ support warnings
* ✅ apply random query filtering on the fly to save memory
* ✅ let amount of work depend on specified execution time
* ✅ support fail-fast on first error
* ✅ support functions with variable parameter count

#### Docs
* ✅ document classes

### V2
(moved to ticket)

### Implementation

UML diagram

### CODE REVIEW

File: misc/python/materialize/output_consistency/data_generation/generator.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+            for new_combination in new_combinations:
+                entry: list[Expression] = [current_value]
+                entry.extend(new_combination)
+                value_combinations.append(entry)
+
+    return value_combinations
+
+
+def is_valid_combination(
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/data_generation/generator.py:85]: Planned to add operation-specific characterstics to determine further combinations resulting in an error (e.g., functions annotated with `VALUE_INCREASES` will fail if one operand has characteristic `MAX_VALUE` and at least one other has no characteristic like `ZERO_OR_NULL`.

File: misc/python/materialize/output_consistency/data_generation/generator.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                data_type.raw_values,
+                offset_value=data_type.raw_values[0],
+                length=operation_type.op_count,
+                with_self=True,
+                with_earlier=not operation_type.commutative,
+            )
+
+            for combination in combinations:
+                if not is_valid_combination(operation_type, combination):
+                    continue
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/data_generation/generator.py:39]: Excluding these combinations for now. Planned to run them in a separate query expecting errors as outcome.

File: misc/python/materialize/output_consistency/data_generation/generator.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            combinations = get_combinations(
+                data_type.raw_values,
+                offset_value=data_type.raw_values[0],
+                length=operation_type.op_count,
+                with_self=True,
+                with_earlier=not operation_type.commutative,
+            )
+
+            for combination in combinations:
+                if not is_valid_combination(operation_type, combination):
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/data_generation/generator.py:38]: Maybe to be renamed to sth like `is_combination_expecting_error`

File: misc/python/materialize/output_consistency/data_generation/picker.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import random
+
+from materialize.output_consistency.expressions.expression import ExpressionWithNArgs
+
+
+def random_seed() -> int:
+    return 0
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/data_generation/picker.py:16]: TODO

File: misc/python/materialize/output_consistency/data_generation/picker.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import random
+
+from materialize.output_consistency.expressions.expression import ExpressionWithNArgs
+
+
+def random_seed() -> int:
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/data_generation/picker.py:15]: General idea: We could run this test with one or multiple _fixed seeds_ to achieve deterministic executions in non-nightly executions. We could run this test _additionally with random seeds_ in nightly builds.

File: misc/python/materialize/output_consistency/execution/execution_strategy.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class EvaluationStrategy:
+    def __init__(self, key: str, name: str):
+        self.key = key
+        self.db_object_name = key
+        self.name = name
+
+    def generate_source(self, data_types: list[DataType]) -> list[str]:
+        raise RuntimeError("Not implemented")
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/execution/execution_strategy.py:20]: TODO: Consider using `@abstractmethod` instead

File: misc/python/materialize/output_consistency/operations/operation_provider.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "$ - $",
+    set(),
+    set(),
+)
+mult_op = OperationWithTwoArgs("$ * $", set(), set(), commutative=True)
+div_op = OperationWithTwoArgs("$ / $", set(), {ValueCharacteristics.ZERO})
+
+sum_func = UnaryFunction("SUM", set())
+
+OPERATION_TYPES: list[OperationWithNArgs] = [
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/operations/operation_provider.py:28]: To be extended

File: misc/python/materialize/output_consistency/values/data_provider.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    "INT2", is_signed=True, is_decimal=False, tiny_value="1", max_value="32767"
+)
+int4_type = NumberDataType(
+    "UINT2", is_signed=True, is_decimal=False, tiny_value="1", max_value="2147483647"
+)
+uint2_type = NumberDataType(
+    "UINT2", is_signed=False, is_decimal=False, tiny_value="1", max_value="65535"
+)
+
+DATA_TYPES = [int2_type, int4_type, uint2_type]
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/values/data_provider.py:25]: To be extended

File: test/output-consistency/mzcompose.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+        c.sql(
+            """
+            SET cluster = cluster1;
+            -- now let's give it another go with user-defined objects
+            CREATE TABLE t (a int);
+            CREATE DEFAULT INDEX ON t;
+            INSERT INTO t VALUES (42);
+            """
+        )
```
Comment (nrainer-materialize) [inline test/output-consistency/mzcompose.py:74]: TODO: cleanup and remove unrelated stuff

File: test/output-consistency/mzcompose.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+        cursor: Cursor = c.sql_cursor()
+        cursor.execute("SET cluster = cluster1;")
+
+        cursor.execute("BEGIN;")
```
Comment (nrainer-materialize) [inline test/output-consistency/mzcompose.py:79]: TODO: run and evaluate generated queries

File: test/output-consistency/mzcompose.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            """
+            SET cluster = cluster1;
+            -- now let's give it another go with user-defined objects
+            CREATE TABLE t (a int);
+            CREATE DEFAULT INDEX ON t;
+            INSERT INTO t VALUES (42);
+            """
+        )
+
+        cursor: Cursor = c.sql_cursor()
```
Comment (nrainer-materialize) [inline test/output-consistency/mzcompose.py:76]: TODO: figure out how to connect to the generation code

**Comment by philip-stoev** [comment]
@nrainer-materialize I will review this on the plane Sunday most likely. Apologies for the delay.

**Comment by nrainer-materialize** [comment]
> @nrainer-materialize I will review this on the plane Sunday most likely. Apologies for the delay.

No worries, that's absolutely fine.

Please be aware that this pull request is a draft. I am not yet looking for a detailed review but **more interested in general feedback about concept & structure.** The implementation as such still contains several TODOs and know issues, and will need further testing.

> @nrainer-materialize in the current state of this PR, is there a way for me to easily print the statements that are generated?

As mentioned above, the implementation is not complete and mature yet. Therefore, I consider it too early to assess the generated statements. I will get back to your question later.

File: test/output-consistency/mzcompose.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from materialize.mzcompose import Composition, WorkflowArgumentParser
+from materialize.mzcompose.services import (
+    Clusterd,
+    Cockroach,
+    Kafka,
```
Comment (def-) [inline test/output-consistency/mzcompose.py:14]: I think Kafka etc currently not used, can be slimmed down?

File: test/output-consistency/mzcompose.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+SERVICES = [
+    Zookeeper(),
+    Kafka(),
+    SchemaRegistry(),
+    Localstack(),
+    Cockroach(setup_materialize=True),
+    Clusterd(name="clusterd1"),
+    # We use mz_panic() in some test scenarios, so environmentd must stay up.
+    Materialized(propagate_crashes=False, external_cockroach=True),
+    Testdrive(
```
Comment (def-) [inline test/output-consistency/mzcompose.py:34]: Testdrive unused too I guess

File: test/output-consistency/mzcompose.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    """
+
+    c.down(destroy_volumes=True)
+    with c.override(
+        Testdrive(no_reset=True),
+    ):
+        c.up("testdrive", persistent=True)
+        c.up("materialized")
+
+    run_output_consistency_tests(c)
```
Comment (def-) [inline test/output-consistency/mzcompose.py:61]: Also would be nice to have an additional mode of running this from command line and passing a postgresql connection string, without `mzcompose` (or maybe that's just me who prefers to run things like that?)

**Review by def- — COMMENTED** [review]
First time seeing UML since my early studies. ;) What generator did you use?

I think would be most interesting to see a sample run with the queries printed out to judge if this is going in the right direction. Since we are not using SLT/TD there will be no easy reproducer printed out but I think it would be valuable to add all the commands so you can easily reproduce an issue with `psql postgres://materialize@localhost:6875 -f file.sql`.

File: test/output-consistency/mzcompose.py (hunk 14 above)
Comment (philip-stoev) [reply test/output-consistency/mzcompose.py:61]: We need both and they are not mutually exclusive.

- writing things mzcompose-first means that it will be easier for them to go beyond the "this is a one-off script" and rapidly move to be included in the CI -- after all, once there is a mzcompose, it is only a very small step to add it to the Nightly CI. So, in this sense, using mzcompose has a disciplining , "ship-it" type of effect.
- at the same time, you can totally have an `--materialize-url` option to the mzcompose workflow. If set, it should not do the `c.up` and move directly to creating a `cursor` object out of the passed URL 

This comment of mine should be worth at least 20 eur. This is how much I paid for Lufthansa in-flight internet and it is so slow that I can not really work on the rest of the PR right now.

File: misc/python/materialize/output_consistency/expressions/expression_with_args.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+            raise RuntimeError(
+                f"Expected {operation.param_count} arguments but got {len(args)}"
+            )
+
+    def to_sql(self) -> str:
+        sql: str = self.pattern
+
+        for arg in self.args:
+            sql = sql.replace("$", arg.to_sql(), 1)
+
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/expressions/expression_with_args.py:36]: Assert that all `$` have been replaced.

File: misc/python/materialize/output_consistency/operations/operation.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        args_validators: Optional[set[OperationArgsValidator]] = None,
+        commutative: bool = False,
+    ):
+        self.pattern = pattern
+        super().__init__(
+            pattern, [param1, param2], return_type_group, args_validators, commutative
+        )
+
+
+class UnaryFunction(OperationWithOneParam):
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/operations/operation.py:66]: How is `UnaryFunction` different from `OperationWithOneParam`? If they are practically the same, consider having just a single class with `Function` in its name.

File: misc/python/materialize/output_consistency/operations/operation_args_validator.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+)
+from materialize.output_consistency.expressions.expression import Expression
+
+
+class OperationArgsValidator:
+    def is_expected_to_cause_error(self, args: list[Expression]) -> bool:
+        raise RuntimeError("Not implemented")
+
+
+class ValueGrowsArgsValidator(OperationArgsValidator):
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/operations/operation_args_validator.py:23]: a `"""` comment describing this would be appreciated.

File: misc/python/materialize/output_consistency/expressions/expression_generator.py [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+from materialize.output_consistency.expressions.expression import Expression
+from materialize.output_consistency.expressions.expression_with_args import (
+    ExpressionWithNArgs,
+)
+from materialize.output_consistency.operations.operation import OperationWithNParams
+from materialize.output_consistency.operations.operation_provider import OPERATION_TYPES
+
+
+class ExpressionGenerator:
+    def generate_expressions(self) -> list[Expression]:
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/expressions/expression_generator.py:21]: So you generate all the expressions at once and then possibly pick N of them using `random.choice()` . This may cause an OOM at some point once you add all the stuff that needs testing. You may need to generate just N random expressions rather than generate all and then throw away most of them.

**Comment by philip-stoev** [comment]
I think this is going on the right track. A couple of general comments:

- please make sure that each class hierarchy has at least some comment describing what it does.

- For aggregates, joins and Filters, a different shape of the data is required -- all values must be present in a single column:
```
CREATE TABLE t1 (f1 INT);
INSERT INTO t1 VALUES (NULL), (0::INT4), (1::INT4), (2147483647::INT4), (-2147483647::INT4);
SELECT f1 / 0 FROM t1;
SELECT MAX(f1) - MIN(f1) FROM t1;
...
```

would it be possible to get this to happen within your existing framework, in addition to the format you already have? I think your class hierarchy should allow those queries to be generated based on the data you aleady are calculating?

File: misc/python/materialize/output_consistency/expressions/expression_generator.py [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+                    if self.is_expected_to_cause_error(operation, combination):
+                        # exclude for now, handle separately later
+                        continue
+
+                    expression = ExpressionWithNArgs(operation, args=combination)
+                    expressions.append(expression)
+
+        return expressions
+
+    def generate_combinations(
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/expressions/expression_generator.py:52]: This method deserves a brief comment describing the algorithm.

**Comment by nrainer-materialize** [comment]
> For aggregates, joins and Filters, a different shape of the data is required -- all values must be present in a single column:
> [...]

As discussed with @philip-stoev, use two different formats for the input data:
* Use multi-column structure for arithmetic operations like `SELECT v1 + v2 FROM t1;`
* Use multi-row structure for scalar operations like `SELECT MAX(v_data_type) FROM t1;`
* Use multi-row structure (and more values) to test push-down-to-persistence: `SELECT v_data_type FROM t1 WHERE v_data_type > 5;`

For multi-row structure, in particular to test the push-down, use multiple insert statements instead of a single insert statement with multiple rows.

File: misc/python/materialize/output_consistency/data_type/data_type_group.py [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+from enum import Enum
+
+
+class DataTypeGroup(Enum):
```
[the remark this answers is not in the document]
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/data_type/data_type_group.py:12]: I am considering renaming it to `DataClass`. In addition, I will add a comment to each class.

File: misc/python/materialize/output_consistency/validation/result_comparator.py [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+                    sql2=failure2.sql,
+                )
+            )
+
+    def normalize_error_message(self, error_message: str) -> str:
+        # replace source prefix in column
+        normalized_message = re.sub(
+            'column "[^.]*\\.', 'column "<source>.', error_message
+        )
+        normalized_message = normalized_message.replace("Evaluation error: ", "")
```
Comment (nrainer-materialize) [inline misc/python/materialize/output_consistency/validation/result_comparator.py:172]: One of the two evaluation strategies uses this prefix, is this a relevant deviation that deserves a ticket?

Comment (def-) [reply misc/python/materialize/output_consistency/validation/result_comparator.py:172]: I think this is expected and is used to wrap errors which happen inside some evaluation context

**Comment by nrainer-materialize** [comment]
V1: Ready for review

File: bin/output-consistency [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+#
+# output-consistency - Test the output consistency of different query evaluation strategies (e.g., dataflow rendering and constant folding).
```
Comment (def-) [inline bin/output-consistency:12]: Maybe the file name should indicate that this is a test

File: ci/nightly/pipeline.yml [hunk 23]
```diff
@@ -590,6 +591,17 @@ steps:
...
+  - id: output-consistency
+    label: "Output consistency"
+    timeout_in_minutes: 58
+    agents:
+      queue: linux-x86_64
+    artifact_paths: junit_*.xml
+    plugins:
+      - ./ci/plugins/mzcompose:
+          composition: output-consistency
+          args: ["--seed=$BUILDKITE_JOB_ID", "--runtime-in-sec=3300", "--dry-run=False", "--fail-fast=False", "--execute-setup=True"]
```
Comment (def-) [inline ci/nightly/pipeline.yml:603]: For the boolean parameters I think `--no-dry-run`, `--no-fail-fast`, `--execute-setup` seem cleaner. But also, aren't those the defaults?

File: misc/python/materialize/output_consistency/execution/query_execution_manager.py [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+    ValidationOutcome,
+)
+
+
+class QueryExecutionManager:
+    """Requests the execution of queries and handles transactions"""
+
+    def __init__(
+        self,
+        evaluation_strategies: list[EvaluationStrategy],
```
Comment (def-) [inline misc/python/materialize/output_consistency/execution/query_execution_manager.py:46]: We usually use `List` via `from typing import List` for type hints. I think it mostly matters for old Python version compatibility.

File: misc/python/materialize/output_consistency/output_consistency.py [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+) -> argparse.Namespace:
+    parser.add_argument("--seed", default=0, type=int)
+    parser.add_argument(
+        "--dry-run", default=False, type=bool, action=argparse.BooleanOptionalAction
+    )
+    parser.add_argument(
+        "--fail-fast", default=False, type=bool, action=argparse.BooleanOptionalAction
+    )
+    parser.add_argument(
+        "--execute-setup",
```
Comment (def-) [inline misc/python/materialize/output_consistency/output_consistency.py:64]: Can we have this also reset the existing data structures?

File: misc/python/materialize/output_consistency/output_consistency.py [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+        type=bool,
+        action=argparse.BooleanOptionalAction,
+    )
+    parser.add_argument(
+        "--verbose",
+        default=False,
+        type=bool,
+        action=argparse.BooleanOptionalAction,
+    )
+    parser.add_argument("--runtime-in-sec", default=600, type=int)
```
Comment (def-) [inline misc/python/materialize/output_consistency/output_consistency.py:75]: Could we have 0 mean unlimited?

File: test/output-consistency/mzcompose.py [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+from materialize.mzcompose.services import Cockroach, Materialized
+from materialize.output_consistency.output_consistency import (
+    parse_output_consistency_input_args,
+    run_output_consistency_tests,
+)
+
+SERVICES = [
+    Cockroach(setup_materialize=True),
+    # We use mz_panic() in some test scenarios, so environmentd must stay up.
+    Materialized(propagate_crashes=False, external_cockroach=True),
```
Comment (def-) [inline test/output-consistency/mzcompose.py:20]: I think you're not using mz_panic here, so best to propagate crashes, then we can also use the ci-logging scripting to autodetect them.

File: misc/python/materialize/output_consistency/validation/validation_outcome.py [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+        return self._problem_marker_output(self.warnings)
+
+    def remark_output(self) -> str:
+        return self._problem_marker_output(self.remarks)
+
+    def _problem_marker_output(self, entries: Sequence[ValidationMessage]) -> str:
+        if len(entries) == 0:
+            return ""
+
+        return LI_PREFIX + f"\n{LI_PREFIX}\n".join([str(entry) for entry in entries])
```
Comment (def-) [inline misc/python/materialize/output_consistency/validation/validation_outcome.py:63]:
```suggestion
        return "\n".join(f"{LI_PREFIX}\n{entry}" for entry in entries)
```
(untested)

**Review by def- — CHANGES_REQUESTED** [review]
I triggered a nightly: https://buildkite.com/materialize/nightlies/builds/2399 which failed: https://buildkite.com/materialize/nightlies/builds/2399#01882bfc-12e0-4fcf-9034-2aa2aec52d25 You can also trigger nightly yourself to try it out.

Also trying it out locally now.

File: ci/nightly/pipeline.yml (hunk 23 above)
Comment (nrainer-materialize) [reply ci/nightly/pipeline.yml]: ✅ That's true, removed.

File: misc/python/materialize/output_consistency/execution/query_execution_manager.py (hunk 24 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/execution/query_execution_manager.py]: ✅ Replaced. Done as well for `set` and `dict`.

File: misc/python/materialize/output_consistency/output_consistency.py (hunk 25 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output_consistency.py:64]: Sure, will do

File: misc/python/materialize/output_consistency/output_consistency.py (hunk 26 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output_consistency.py:75]: Sure

File: test/output-consistency/mzcompose.py (hunk 27 above)
Comment (nrainer-materialize) [reply test/output-consistency/mzcompose.py]: ✅

File: misc/python/materialize/output_consistency/output_consistency.py (hunk 26 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output_consistency.py:80]: ✅

File: misc/python/materialize/output_consistency/output_consistency.py (hunk 25 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output_consistency.py:68]: ✅

**Comment by def-** [comment]
Looks like nightly is finding quite some issues: https://buildkite.com/materialize/nightlies/builds/2444#018844d5-71bc-4c48-8ec2-cffc8f02ca9c Are those known already and can they be ignored?

**Comment by nrainer-materialize** [comment]
> Looks like nightly is finding quite some issues: https://buildkite.com/materialize/nightlies/builds/2444#018844d5-71bc-4c48-8ec2-cffc8f02ca9c Are those known already and can they be ignored?

I am working on those, I am expecting them all to have the same root cause: apparently, two different `NaN` values are delivered.

File: ci/nightly/pipeline.yml [hunk 29]
```diff
@@ -590,6 +591,17 @@ steps:
...
+  - id: output-consistency
+    label: "Output consistency"
+    timeout_in_minutes: 58
+    agents:
+      queue: linux-x86_64
+    artifact_paths: junit_*.xml
+    plugins:
+      - ./ci/plugins/mzcompose:
+          composition: output-consistency
+          args: ["--seed=$BUILDKITE_JOB_ID", "--max-runtime-in-sec=180"]
```
Comment (philip-stoev) [inline ci/nightly/pipeline.yml:603]: Why did you decide on a 3-minute run in Nightly? I would say 10-30 min is more in line with the rest of the fuzzing workloads that are in Nightly (unless you consider the marginal utility from longer runs to be very minimal).

File: misc/python/materialize/output_consistency/common/format_constants.py [hunk 30]
```diff
@@ -0,0 +1,15 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+CONTENT_SEPARATOR_1 = "==============="
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/common/format_constants.py:10]: More descriptive names of the constantsmay help here.

File: misc/python/materialize/output_consistency/execution/sql_executor.py [hunk 31]
```diff
@@ -0,0 +1,122 @@
...
+
+        if details is None:
+            return f"{message}"
+        else:
+            return f"{message} ({details})"
+
+
+class DryRunSqlExecutor(SqlExecutor):
+    def consume_sql(self, sql: str) -> None:
+        print(f"> {sql}")
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/execution/sql_executor.py:97]: Right now the output of the dry run is mostly non-testdrive-compatible, so printing just the query in testdrive format makes copy+paste difficult. Consider just dropping the "> here".

The output of dry-run does not form a stand-alone SQL script on its own, since it has non-commented-out freeform print lines, but not having  the ">" should make it easier to copy the output over to some other place (github ticket, etc.)

File: misc/python/materialize/output_consistency/execution/test_summary.py [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+
+    def __init__(
+        self,
+        dry_run: bool = False,
+        count_executed_query_templates: int = 0,
+        count_successful_query_templates: int = 0,
+        count_with_warning_query_templates: int = 0,
+    ):
+        super().__init__()
+        self.mode = "DB" if not dry_run else "DRY_RUN"
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/execution/test_summary.py:33]: consider `DB` -> `LIVE_DATABASE` ?

File: test/output-consistency/mzcompose.py [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+
+    c.down(destroy_volumes=True)
+
+    c.up("materialized")
+
+    args = parse_output_consistency_input_args(parser)
+    connection = c.sql_connection()
+
+    test_summary = run_output_consistency_tests(connection, args)
+
```
Comment (philip-stoev) [inline test/output-consistency/mzcompose.py:36]: Maybe print the summary here in a `finally` block?

File: misc/python/materialize/output_consistency/input_data/types/number_types_provider.py [hunk 34]
```diff
@@ -0,0 +1,152 @@
...
+    "DECIMAL(39)",
+    is_signed=True,
+    is_decimal=True,
+    tiny_value="1",
+    max_value="999999999999999999999999999999999999999",
+    max_negative_value="-999999999999999999999999999999999999999",
+)
+DECIMAL39_4_TYPE = NumberDataType(
+    "DECIMAL_39_4",
+    "DECIMAL(39,4)",
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/input_data/types/number_types_provider.py:103]: Consider giving a scale of 8 instead of 4, to catch more obscure rounding/casting errors.

File: misc/python/materialize/output_consistency/expression/expression_characteristics.py [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+class ExpressionCharacteristics(Enum):
+    """Characteristics describing an `Expression`"""
+
+    NULL = 0
+    # not (NULL or ZERO)
+    NON_EMPTY = 1
+
+    ZERO = 100
+    ONE = 101
+    TINY_VALUE = 102
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/expression/expression_characteristics.py:22]: As a future enhancement, also consider values that are:
- just below 0.5
- just above 0.5
- just below 1

For the purpose of sussing out any issues with rounding, ceiling, fooring and casting.

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py [hunk 36]
```diff
@@ -0,0 +1,39 @@
...
+)
+from materialize.output_consistency.operation.operation_param import OperationParam
+
+
+class NumericOperationParam(OperationParam):
+    def __init__(
+        self,
+        optional: bool = False,
+        incompatibilities: Optional[Set[ExpressionCharacteristics]] = None,
+        incompatibility_combinations: Optional[
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:24]: The concept of incompatibility_combinations may benefit from a comment

File: misc/python/materialize/output_consistency/input_data/values/number_values_provider.py [hunk 37]
```diff
@@ -0,0 +1,113 @@
...
+    if type_definition.is_decimal:
+        values_of_type.add_characteristic_to_all_values(
+            ExpressionCharacteristics.DECIMAL_OR_FLOAT_TYPED
+        )
+
+    for value in values_of_type.raw_values:
+        if ExpressionCharacteristics.MAX_VALUE in value.characteristics:
+            value.characteristics.add(ExpressionCharacteristics.LARGE_VALUE)
+
+VALUES_PER_NUMERIC_DATA_TYPE[UINT4_TYPE].add_characteristic_to_all_values(
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/input_data/values/number_values_provider.py:83]: Are all values `LARGER_THAN_INT4_SIZED` or just the `MAX` ?

File: misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py [hunk 38]
```diff
@@ -0,0 +1,461 @@
...
+NUMERIC_OPERATION_TYPES.append(
+    DbOperation(
+        "$ * $",
+        [NumericOperationParam(), NumericOperationParam()],
+        DataTypeCategory.NUMERIC,
+        {MultiParamValueGrowsArgsValidator()},
+    )
+)
+NUMERIC_OPERATION_TYPES.append(
+    DbOperation(
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py:59]: Division and mod should all be high-relevance.

File: misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py [hunk 39]
```diff
@@ -0,0 +1,461 @@
...
+        "$ % $",
+        [
+            NumericOperationParam(),
+            NumericOperationParam(incompatibilities={ExpressionCharacteristics.ZERO}),
+        ],
+        DataTypeCategory.NUMERIC,
+    )
+)
+# Bitwise AND
```
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py:78]: All the bitwise operators are low relevance I think.

File: misc/python/materialize/output_consistency/execution/sql_executor.py (hunk 31 above)
Comment (philip-stoev) [reply misc/python/materialize/output_consistency/execution/sql_executor.py:97]: Also, is it possible to print that a given query is expecting an error during the dry run? That would showcase the "expects error" mechanism.

**Review by philip-stoev — APPROVED** [review]
Thank you, that is quite the body of code.

Would you care to add the rest of a database implementation to that type checker of yours? By May 31st? I will get some investors lined up.

File: ci/nightly/pipeline.yml (hunk 29 above)
Comment (nrainer-materialize) [reply ci/nightly/pipeline.yml]: ✅ Changed to 20 mins. I was surprised how fast it is (at least compared to what I had expected 😀). However, it is way faster locally than in the CI.

File: misc/python/materialize/output_consistency/common/format_constants.py (hunk 30 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/common/format_constants.py]: ✅ Renamed

File: misc/python/materialize/output_consistency/execution/test_summary.py (hunk 32 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/execution/test_summary.py]: ✅

File: test/output-consistency/mzcompose.py (hunk 33 above)
Comment (nrainer-materialize) [reply test/output-consistency/mzcompose.py:36]: I think this is the only comment I haven't addressed...

What would you like to achieve by that? The summary is printed from within `run_output_consistency_tests` (or somewhere deeper). Just adding a finally block will not work because `test_summary` won't be in scope there.

File: misc/python/materialize/output_consistency/input_data/types/number_types_provider.py (hunk 34 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/types/number_types_provider.py]: ✅ https://github.com/MaterializeInc/materialize/pull/19013/commits/76c00c8d47712b045798d4f420af3111fdb41c62

File: misc/python/materialize/output_consistency/expression/expression_characteristics.py (hunk 35 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/expression/expression_characteristics.py:22]: ✅ https://github.com/MaterializeInc/materialize/pull/19013/commits/5a5400e379e85caeda0da92341074b8b706bd893

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py (hunk 36 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:24]: ✅

File: misc/python/materialize/output_consistency/input_data/values/number_values_provider.py (hunk 37 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/values/number_values_provider.py:95]: It is a bit misleading, I renamed it to `TYPE_LARGER_INT4_SIZED`. Although assigned to the individual value, it is about the type.

File: misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py (hunk 38 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py:59]: ✅

File: misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py (hunk 39 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/operations/number_operations_provider.py:80]: ✅

**Comment by nrainer-materialize** [comment]
> First time seeing UML since my early studies. ;) What generator did you use?
> 
> I think would be most interesting to see a sample run with the queries printed out to judge if this is going in the right direction. Since we are not using SLT/TD there will be no easy reproducer printed out but I think it would be valuable to add all the commands so you can easily reproduce an issue with `psql postgres://materialize@localhost:6875 -f file.sql`.

The UML tool is called umlet. It is pretty basic but I like it.

**Comment by nrainer-materialize** [comment]
✅ Nightly is green: https://buildkite.com/materialize/nightlies/builds/2457#01884902-f1b9-4657-9c33-998104681b18
